File: frontend/src/api/aggregator.ts
//import { interceptors } from "undici-types";

const API_URL = process.env.NEXT_PUBLIC_API_URL || "http://localhost:8080";

// Terraform 출력 타입 정의
export interface TerraformOutput {
  instanceId?: string;
  publicIp?: string;
  privateIp?: string;
  region?: string;
  [key: string]: unknown;
}

// 실제 API 응답 구조에 맞는 Aggregator 정보 타입 정의
export interface AggregatorInfo {
  id: string;
  user_id: number;
  name: string;
  status: string;
  algorithm: string;
  cloud_provider: string;
  project_name: string;
  region: string;
  zone: string;
  instance_type: string;
  participant_count: number;
  current_round: number;
  current_cost: number;
  estimated_cost: number;
  cpu_specs: string;
  memory_specs: string;
  storage_specs: string;
  cpu_usage: number;
  memory_usage: number;
  network_usage: number;
  instance_id: string;
  public_ip: string;
  private_ip: string;
  created_at: string;
  updated_at: string;
  user: {
    id: number;
    email: string;
    name: string;
    created_at: string;
    updated_at: string;
  };
  federated_learning?: {
    id: string;
    user_id: number;
    cloud_connection_id: string;
    aggregator_id: string;
    name: string;
    description: string;
    status: string;
    participant_count: number;
    completed_at: string | null;
    accuracy: string;
    rounds: number;
    algorithm: string;
    model_type: string;
    created_at: string;
    updated_at: string;
  };
  terraformOutput?: TerraformOutput; // 기존 필드 유지
}

// Aggregator 배치 최적화설정 타입 정의
export interface AggregatorOptimizeConfig {
  maxBudget: number;
  maxLatency: number;
  weightBalance?: number;
}

export interface AggregatorConfig {
  cloudProvider: string;
  region: string;
  instanceType: string;
  memory: number;
  projectId?: string; // GCP용 프로젝트 ID (선택적)
  estimatedCost: number;
}

// 연합학습 데이터 타입 정의 (실제 API 구조에 맞게 수정)
export interface FederatedLearningData {
  id: string;
  user_id: number;
  cloud_connection_id: string;
  aggregator_id: string;
  name: string;
  description: string;
  status: string;
  participant_count: number;
  completed_at: string | null;
  accuracy: string;
  rounds: number;
  algorithm: string;
  model_type: string;
  created_at: string;
  updated_at: string;
  participants?: Array<{
    id: string;
    name: string;
    status: string;
    openstack_endpoint?: string;
  }>;
  modelFileName?: string | null;
  modelFileSize?: number; // 모델 파일 크기 (바이트 단위)
}

// Aggregator 생성 요청 타입
export interface CreateAggregatorRequest {
  name: string;
  algorithm: string;
  region: string;
  storage: string; // GB as string
  instanceType: string;
  cloudProvider: string; // "aws" | "gcp"
  projectId?: string; // GCP용 프로젝트 ID (선택적)
  estimatedCost: string;
}

// Aggregator 생성 응답 타입
export interface CreateAggregatorResponse {
  aggregatorId: string;
  status: string;
  terraformStatus?: string;
}

// Aggregator 제어 응답 타입
export interface AggregatorControlResponse {
  status: string;
  message?: string;
  data?: JSON;
}

// API 응답에서 사용되는 상태 값들
export type AggregatorStatus =
  | "running"
  | "completed"
  | "error"
  | "pending"
  | "creating"
  | "failed"
  | "ready"
  | "stopped";

export type FederatedLearningStatus =
  | "ready"
  | "running"
  | "completed"
  | "failed"
  | "pending";

// 클라우드 제공자 타입
export type CloudProvider = "aws" | "gcp" | "azure";

// 모델 타입
export type ModelType = "nlp" | "cv" | "tabular" | "time_series";

// Aggregator 생성 함수
export const createAggregator = async (
  federatedLearningData: FederatedLearningData,
  aggregatorConfig: AggregatorConfig
): Promise<CreateAggregatorResponse> => {
  try {
    // Derive simple storage size in GB (fallback to 20GB)
    const derivedStorageGB = Math.max(
      20,
      Math.ceil(
        (federatedLearningData.modelFileSize || 0) / (1024 * 1024 * 1024) + 5
      )
    );

    // 중복 방지를 위해 타임스탬프 추가
    const timestamp = new Date()
      .toISOString()
      .replace(/[:.]/g, "-")
      .slice(0, -5);
    const uniqueName = `${federatedLearningData.name}-${timestamp}`;

    const requestBody: CreateAggregatorRequest = {
      name: uniqueName,
      algorithm: federatedLearningData.algorithm,
      region: aggregatorConfig.region,
      storage: String(derivedStorageGB),
      instanceType: aggregatorConfig.instanceType,
      cloudProvider: aggregatorConfig.cloudProvider || "aws", // 기본값 AWS
      projectId: aggregatorConfig.projectId, // GCP용
      estimatedCost: String(aggregatorConfig.estimatedCost),
    };

    const response = await fetch(`${API_URL}/api/aggregators`, {
      method: "POST",
      credentials: "include",
      headers: {
        "Content-Type": "application/json",
      },
      body: JSON.stringify(requestBody),
    });

    if (!response.ok) {
      const errorData = await response.json().catch(() => ({}));
      console.error("API 호출 실패:", errorData);

      // 중복 생성 에러에 대한 특별한 처리
      if (
        response.status === 400 &&
        errorData.error &&
        errorData.error.includes("동일한 이름의 집계자가 이미 존재합니다")
      ) {
        throw new Error(
          "동일한 이름의 집계자가 이미 존재합니다. 다른 이름을 사용해주세요."
        );
      }

      throw new Error(
        errorData.error || `HTTP error! status: ${response.status}`
      );
    }

    const data = await response.json();
    console.log("API 응답 성공:", data);
    return data.data;
  } catch (error) {
    console.error("Aggregator 생성에 실패했습니다:", error);
    throw error;
  }
};

// Aggregator 상태 조회 함수
export const getAggregatorStatus = async (
  aggregatorId: string
): Promise<{
  status: string;
  progress?: number;
  message?: string;
  terraformStatus?: string;
}> => {
  try {
    // Backend exposes GET /api/aggregators/{id}
    const response = await fetch(`${API_URL}/api/aggregators/${aggregatorId}`, {
      credentials: "include",
      headers: {
        "Content-Type": "application/json",
      },
    });

    if (!response.ok) {
      throw new Error(`HTTP error! status: ${response.status}`);
    }

    const data = await response.json();
    const agg = data?.data ?? data;
    return {
      status: agg.status,
      terraformStatus: agg.terraformStatus,
    };
  } catch (error) {
    console.error("Aggregator 상태 조회에 실패했습니다:", error);
    throw error;
  }
};

// Aggregator 목록 조회 함수
export const getAggregators = async (): Promise<AggregatorInfo[]> => {
  try {
    const response = await fetch(`${API_URL}/api/aggregators`, {
      credentials: "include",
      headers: {
        "Content-Type": "application/json",
      },
    });

    if (!response.ok) {
      throw new Error(`HTTP error! status: ${response.status}`);
    }

    const data = await response.json();
    return Array.isArray(data) ? data : data.data;
  } catch (error) {
    console.error("Aggregator 목록 조회에 실패했습니다:", error);
    throw error;
  }
};

// Aggregator 삭제 함수
export const deleteAggregator = async (aggregatorId: string): Promise<void> => {
  try {
    const response = await fetch(`${API_URL}/api/aggregators/${aggregatorId}`, {
      method: "DELETE",
      credentials: "include",
      headers: {
        "Content-Type": "application/json",
      },
    });

    if (!response.ok) {
      const errorData = await response.json().catch(() => ({}));
      throw new Error(
        errorData.error || `HTTP error! status: ${response.status}`
      );
    }
  } catch (error) {
    console.error("Aggregator 삭제에 실패했습니다:", error);
    throw error;
  }
};

// Aggregator 일시정지 함수
export const pauseAggregator = async (
  aggregatorId: string
): Promise<AggregatorControlResponse> => {
  try {
    const response = await fetch(`${API_URL}/api/aggregators/${aggregatorId}`, {
      method: "PUT",
      credentials: "include",
      headers: {
        "Content-Type": "application/json",
      },
      body: JSON.stringify({
        action: "pause",
      }),
    });

    if (!response.ok) {
      const errorData = await response.json().catch(() => ({}));
      throw new Error(
        errorData.error || `HTTP error! status: ${response.status}`
      );
    }

    const data = await response.json();
    console.log(`Aggregator ${aggregatorId} 일시정지 성공:`, data);
    return data;
  } catch (error) {
    console.error(`Aggregator ${aggregatorId} 일시정지에 실패했습니다:`, error);
    throw error;
  }
};

// Aggregator 재개 함수
export const resumeAggregator = async (
  aggregatorId: string
): Promise<AggregatorControlResponse> => {
  try {
    const response = await fetch(`${API_URL}/api/aggregators/${aggregatorId}`, {
      method: "PUT",
      credentials: "include",
      headers: {
        "Content-Type": "application/json",
      },
      body: JSON.stringify({
        action: "resume",
      }),
    });

    if (!response.ok) {
      const errorData = await response.json().catch(() => ({}));
      throw new Error(
        errorData.error || `HTTP error! status: ${response.status}`
      );
    }

    const data = await response.json();
    console.log(`Aggregator ${aggregatorId} 재개 성공:`, data);
    return data;
  } catch (error) {
    console.error(`Aggregator ${aggregatorId} 재개에 실패했습니다:`, error);
    throw error;
  }
};

// 집계자 배치 최적화 응답 타입
export interface AggregatorOption {
  rank: number;
  region: string;
  instanceType: string;
  cloudProvider: string;
  estimatedMonthlyCost: number;
  estimatedHourlyPrice: number;
  avgLatency: number;
  maxLatency: number;
  vcpu: number;
  memory: number;
  recommendationScore: number;
}
export interface OptimizationResponse {
  status: string;
  summary: {
    totalParticipants: number;
    participantRegions: string[];
    totalCandidateOptions: number;
    feasibleOptions: number;
    constraints: {
      maxBudget: number;
      maxLatency: number;
    };
    modelInfo: {
      modelType: string;
      algorithm: string;
      rounds: number;
    };
  };
  optimizedOptions: AggregatorOption[];
  message: string;
}

// 집계자 배치 최적화 함수
export const optimizeAggregatorPlacement = async (
  federatedLearningData: FederatedLearningData,
  constraints: {
    maxBudget: number;
    maxLatency: number;
    minMemoryRequirement?: number;
  }
): Promise<OptimizationResponse> => {
  try {
    const requestBody = {
      federatedLearning: federatedLearningData,
      aggregatorConfig: constraints,
    };

    const response = await fetch(`${API_URL}/api/aggregators/optimization`, {
      method: "POST",
      credentials: "include",
      headers: {
        "Content-Type": "application/json",
      },
      body: JSON.stringify(requestBody),
    });

    if (!response.ok) {
      const errorData = await response.json().catch(() => ({}));
      throw new Error(
        errorData.error || `HTTP error! status: ${response.status}`
      );
    }

    const data = await response.json();
    return data.data;
  } catch (error) {
    console.error("집계자 배치 최적화에 실패했습니다:", error);
    throw error;
  }
};

File: backend/models/aggregator.go
package models

import "time"

type Aggregator struct {
	ID            string `json:"id" gorm:"primaryKey"`
	UserID        int64  `json:"user_id" gorm:"not null;index"`
	Name          string `json:"name" gorm:"not null"`
	Status        string `json:"status" gorm:"default:pending"` // pending, running, completed, error
	Algorithm     string `json:"algorithm" gorm:"not null"`     // FedAvg, FedProx, FedAdam, etc.
	CloudProvider string `json:"cloud_provider" gorm:"not null"`

	// 클라우드 공통 필드
	ProjectName  string `json:"project_name" gorm:"not null"`
	Region       string `json:"region" gorm:"not null"`
	Zone         string `json:"zone" gorm:"not null"`
	InstanceType string `json:"instance_type" gorm:"not null"`

	// GCP 전용 필드 (nullable)
	ProjectID *string `json:"project_id,omitempty"` // GCP에서만 사용

	// MLflow 연동 필드 추가
    MLflowExperimentID   *string `json:"mlflow_experiment_id,omitempty"`
    MLflowExperimentName *string `json:"mlflow_experiment_name,omitempty"`
    MLflowRunID         *string `json:"mlflow_run_id,omitempty"`

	ParticipantCount int      `json:"participant_count" gorm:"default:0"`
	CurrentRound     int      `json:"current_round" gorm:"default:0"`
	Accuracy         *float64 `json:"accuracy,omitempty"`
	CurrentCost      float64  `json:"current_cost" gorm:"default:0"`
	EstimatedCost    float64  `json:"estimated_cost" gorm:"default:0"`
	CPUSpecs         string   `json:"cpu_specs"`
	MemorySpecs      string   `json:"memory_specs"`
	StorageSpecs     string   `json:"storage_specs"`
	CPUUsage         float64  `json:"cpu_usage" gorm:"default:0"`
	MemoryUsage      float64  `json:"memory_usage" gorm:"default:0"`
	NetworkUsage     float64  `json:"network_usage" gorm:"default:0"`
	InstanceID       string   `json:"instance_id,omitempty"`
	PublicIP         string   `json:"public_ip,omitempty"`
	PrivateIP        string   `json:"private_ip,omitempty"`

	CreatedAt time.Time `json:"created_at" gorm:"autoCreateTime"`
	UpdatedAt time.Time `json:"updated_at" gorm:"autoUpdateTime"`

	// Relationships
	User              User               `json:"user,omitempty" gorm:"foreignKey:UserID;constraint:OnUpdate:CASCADE,OnDelete:CASCADE"`
	FederatedLearning *FederatedLearning `json:"federated_learning,omitempty" gorm:"foreignKey:AggregatorID"`
}

func (Aggregator) TableName() string {
	return "aggregators"
}

// AggregatorMetrics는 실시간 메트릭을 위한 구조체입니다
type AggregatorMetrics struct {
	AggregatorID string    `json:"aggregator_id"`
	CPUUsage     float64   `json:"cpu_usage"`
	MemoryUsage  float64   `json:"memory_usage"`
	NetworkUsage float64   `json:"network_usage"`
	Timestamp    time.Time `json:"timestamp"`
}

// TrainingRound는 학습 라운드 정보를 위한 구조체입니다
type TrainingRound struct {
	ID                string      `json:"id" gorm:"primaryKey"`
	AggregatorID      string      `json:"aggregator_id" gorm:"not null;index"`
	Round             int         `json:"round" gorm:"not null"`
	ModelMetrics      ModelMetric `json:"model_metrics" gorm:"embedded;embeddedPrefix:model_metrics_"`
	Duration          int         `json:"duration"`           // seconds
	ParticipantsCount int         `json:"participants_count"` // 왜 있는거지?
	StartedAt         time.Time   `json:"started_at"`
	CompletedAt       *time.Time  `json:"completed_at,omitempty"`
	CreatedAt         time.Time   `json:"created_at" gorm:"autoCreateTime"` // 삭제?

	// Relationships
	Aggregator *Aggregator `json:"aggregator,omitempty" gorm:"foreignKey:AggregatorID;constraint:OnUpdate:CASCADE,OnDelete:CASCADE"`
}

type ModelMetric struct {
	Accuracy  *float64 `json:"accuracy,omitempty"`
	Loss      *float64 `json:"loss,omitempty"`
	Precision *float64 `json:"precision,omitempty"`
	Recall    *float64 `json:"recall,omitempty"`
	F1Score   *float64 `json:"f1_score,omitempty"`
}
